test/monster_test/monster_test.c

```diff
@@ -1945,8 +1945,9 @@ int test_fixed_size_array(flatcc_builder_t *B)
     foobar->tests[0].b = 4;
     foobar->tests[1].a = 1;
     foobar->tests[1].b = 2;
-    strncpy(foobar->text, "hello, world", ns(FooBar_text_get_len()));
-    // or strncopy(foobar->text, "hello, world", sizeof(foobar->text));
+    strncpy(foobar->text, "hello, world", sizeof(foobar->text));
+    foobar->text[sizeof(foobar->text) - 1] = '\0';
+
```

Comment:
We are testing that we are not writing too much and that we use all of the 5 characters available. Hence the zero termination should not be there. I'm not sure how to silence that warning. There are many valid use cases for that syntax. We are also testing that the get_len() function is present and does it's job, so removing it does  no good. Maybe we need to truncate the string to "hello". Why does GCC have to be this way, every year.

Reply:
I tried changing this to "hello", but:
```
error: ‘strncpy’ output truncated before terminating nul copying 5 bytes from a string of the same length [-Werror=stringop-truncation]
 1948 |     strncpy(foobar->text, "hello", sizeof(foobar->text));
```
I'm not really sure what this is testing, so I'm open to any ideas on how to fix it.

Reply:
The test is partially a documention of a feature, partially a test.
As documentation it shows how strncpy fills an array with characters padded with zeroes up to the specified length. Char arrays in FlatBuffers are not required to be zero terminated, but they are required to be zero padded. strncpy implements these semantics perfectly.

In the concrete test we test the if the input is longer (truncation), we get the expected prefix without zero termination. A later part of the test reads back the buffer and checks for the 5 bytes "hello". That is, we can store a string without zero padding.

(Whether users want to use zero termination or not, is up to the user, but in this case it is not dictated by the format because we want the array structure to be able to hold arbitrary data and to be compatible with the uint8 array).

The write part of the test goes:
1. Given a a char array of length 5. 
2. Test that the get_len() function is present and can be used as arg to strncopy.
    (We do not test get_len() == 5, but we could).
3. We test the strncopy copies exactly get_len() bytes into the array. (we can't do this perfectly as the buffer might hold padding, but at least it must not overwrite valid data in the buffer and it must write all data expected.

We document in a comment that sizeof can also be used because the array is indeed implemented as a C-array, but this is not what we are testing.

I don't know how to fix this with GCC 9 because I don't have GCC 9, but I do want this to work with strncpy in general, because it does the right thing for this use case.

I am considering reducing the warning level of GCC because I spend so much time on counterproductive warnings. Yet GCC does not find 32/64 mixed types that Windows MSVC is complaining about as a warning (and which I should fix).

Reply:
Depending on how familiar you are with Flatbuffers: strings are normally stored as separate objects with their own length prefix field and a trailing zero terminator always. Char arrays is new, not yet released feature along with other fixed length arrays in structs. The Google flatbuffers project is still debating whether there should be a char array, or it should be an uint8 with an attribute to mark it for use with string content. Either way, flatcc would end up with the same char array and be compatible with uint8 arrays. Other languages might not have a char array concept and would map this to uint8 arrays regardless - also a reason to not force zero termination.